src/node_buffer.cc

```diff
@@ -310,6 +310,70 @@ void Base64Slice(const FunctionCallbackInfo<Value>& args) {
 }
 
 
+// bytesCopied = buffer.mask(mask, target, targetOffset, sourceStart, sourceEnd)
+void Mask(const FunctionCallbackInfo<Value> &args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  uint32_t mask32 = args[0]->Uint32Value();
+  char* mask = (char*)&mask32;
+
+  Local<Object> target = args[1]->ToObject(env->isolate());
+
+  if (!HasInstance(target))
+    return env->ThrowTypeError("second arg should be a Buffer");
+
+  ARGS_THIS(args.This())
+  size_t target_length = target->GetIndexedPropertiesExternalArrayDataLength();
+  char* target_data = static_cast<char*>(
+      target->GetIndexedPropertiesExternalArrayData());
+  size_t target_start;
+  size_t source_start;
+  size_t source_end;
+
+  CHECK_NOT_OOB(ParseArrayIndex(args[2], 0, &target_start));
+  CHECK_NOT_OOB(ParseArrayIndex(args[3], 0, &source_start));
+  CHECK_NOT_OOB(ParseArrayIndex(args[4], obj_length, &source_end));
+
+  // Copy 0 bytes; we're done
+  if (target_start >= target_length || source_start >= source_end)
+    return args.GetReturnValue().Set(0);
+
+  if (source_start > obj_length)
+    return env->ThrowRangeError("out of range index");
+
+  if (source_end - source_start > target_length - target_start)
+    source_end = source_start + target_length - target_start;
+
+  uint32_t to_copy = MIN(MIN(source_end - source_start,
+                             target_length - target_start),
+                             obj_length - source_start);
+
+  obj_data += source_start;
+  target_data += target_start;
+
+  uint32_t* target_data_32 = (uint32_t*)target_data;
+  uint32_t* obj_data_32 = (uint32_t*)obj_data;
```

Comment:
So, I see two issues here: one is pointer aliasing, the other is endianness.

Endianness: doing 32 bits xor operations on bytes in a buffer will produce different results on BE and LE architectures.

Pointer aliasing: the safe way to go about it is as follows:

```c++
union {
  char bytes[sizeof(uint32_t)];
  uint32_t value;
} u;

for (size_t i = 0; i < to_copy_32; i += sizeof(u.bytes)) {
  char* const pointer = obj_data + i;
  memcpy(u.bytes, pointer, sizeof(u.bytes));
  u.value ^= mask32;
  memcpy(pointer, u.bytes, sizeof(u.bytes));
}
```

Not a paragon of beauty but there you have it.